

Code in this directory is used to process the data generated by scCAT-seq, C1 CAGE, BAT-seq, etc. `Fastq` file is needed and the output file format is `BED`. Each Row represents read in corresponding position.

Then `BED` file as input is needed to call peak using `CAGEr` R package.

Here is the scCAT-seq 5' data and 3' data processing workflow. To see detail imformation of other data processing, please see `C1_CAGE_5_data_processing.sh`, `C1_STRT_5_data_processing.sh`, `Arguel_et_al_5_data_processing.sh` and `BAT-seq_3_data_processing.sh`.


We have uploaded test data. Reader can download at [here](https://drive.google.com/open?id=1t8oLqAIWWy32yf5g3NOfKm10-i0pBITy) and [here](https://drive.google.com/open?id=1Z4xEVmkip3aq56Jp5k-0qBLmZ9oR-Lyk).

data processing and peak correction

---

# 1. Preparation

In this workflow, we assume that the genome fa file is located in `~/index/mm10/mm10.fa`. STAR index is located in `~/index/mm10_STAR/`. fastq files from scCAT-seq are located in `~/fastq/`.

STAR index must be prepared before running this workflow. For build index, you can run:

```
## you can change fa and gtf file path
STAR --runThreadN 24 --runMode genomeGenerate \
--genomeDir ~/index/mm10_STAR/ \
--genomeFastaFiles ~/index/mm10/mm10.fa \
--sjdbGTFfile ~/index/mm10/gencode.vM18.annotation.gtf \
--sjdbOverhang 150
```

# 2. Convert to bed

We use `convert_to_bed.sh` to process the data. The script usage is:

```
sh convert_to_bed.sh <genome fa file> <STAR index> <fastq file dir> <output dir>
```

For example:
```
sh convert_to_bed.sh ~/index/mm10/mm10.fa ~/index/mm10_STAR/ ~/fastq/ ~/scCAT_seq/
```

Output files are stored in `~/scCAT_seq/five_pirme/final_out` and `~/scCAT_seq/three_pirme/final_out`. `*remove_trRNA.bed` are needed for upstream analysis. Its format is:

```
chr1    3214161 3214286 ST-E00126:647:HL57HCCXY:1:2107:20354:17377      255     -
chr1    3539203 3539327 ST-E00126:647:HL57HCCXY:1:1202:30188:47562      255     -
chr1    3539203 3539328 ST-E00126:647:HL57HCCXY:1:2202:9820:48142       255     -
chr1    3540150 3540274 ST-E00126:647:HL57HCCXY:1:2105:32177:70592      255     -
chr1    3540150 3540274 ST-E00126:647:HL57HCCXY:1:1217:19380:22018      255     -
chr1    3612996 3613077 ST-E00126:647:HL57HCCXY:1:1117:10389:26484      255     -
```
This is a tab-delimited file. Each row represents a read.




# 3. Call peak and correction




* Features:
1) TPM_of_peak: The total Tags per Million(TPM) value of the peak called by `CAGEr`.
2) TPM_of_Dominant_Site: The highest TPM value of all sites within a peak.
3) Gene_length: The length of the transcript annotated.
4) Peak_width: The width of the peak called by `CAGEr`.
5) Dominant_TPM_to_Smart2: The ratio of Dominant_TPM to the RPM value of the corresponding gene revealed by Smart-seq2.
6) Slope_smart2_curve: The slope of Smart-seq2 coverage curve around the peaks.
7) Trend_of_smart2_reads: Calculated by dividing the number of reads increased/decreased within 50bp distance by 50.
8) Percentage: The percentage of read counts of a peak to the total counts of a transcript.
9) Motif: Motif information around TSS/TES peak

* Prediction:  
1) Predicted each individual peaks in the input file. The value `1` indicates a true TSS/TES peak while value `0` indicates a false TSS/TES peak






