

Code in this directory is used to process the data generated by scCAT-seq, C1 CAGE, BAT-seq, etc. `Fastq` file is needed and the output file format is `BED`. Each Row represents read in corresponding position.

Then `BED` file as input is needed to call peak using `CAGEr` R package.

Here is the scCAT-seq 5' data and 3' data processing workflow. To see detail imformation of other data processing, please see `C1_CAGE_5_data_processing.sh`, `C1_STRT_5_data_processing.sh`, `Arguel_et_al_5_data_processing.sh` and `BAT-seq_3_data_processing.sh`.


We have uploaded test data. Reader can download at [here](https://drive.google.com/open?id=1t8oLqAIWWy32yf5g3NOfKm10-i0pBITy) and [here](https://drive.google.com/open?id=1Z4xEVmkip3aq56Jp5k-0qBLmZ9oR-Lyk).

data processing and peak correction

---

# 1. Preparation

In this workflow, we assume that the genome fa file is located in `~/index/mm10/mm10.fa`. STAR index is located in `~/index/mm10_STAR/`. fastq files from scCAT-seq are located in `~/fastq/`.

STAR index must be prepared before running this workflow. For build index, you can run:

```
## you can change fa and gtf file path
STAR --runThreadN 24 --runMode genomeGenerate \
--genomeDir ~/index/mm10_STAR/ \
--genomeFastaFiles ~/index/mm10/mm10.fa \
--sjdbGTFfile ~/index/mm10/gencode.vM18.annotation.gtf \
--sjdbOverhang 150
```

# 2. Convert to bed

We use `convert_to_bed.sh` to process the data. The script usage is:

```
sh convert_to_bed.sh <genome fa file> <STAR index> <fastq file dir> <output dir>
```

For example:
```
sh convert_to_bed.sh ~/index/mm10/mm10.fa ~/index/mm10_STAR/ ~/fastq/ ~/scCAT_seq/
```

Output files are stored in `~/scCAT_seq/five_pirme/final_out` and `~/scCAT_seq/three_pirme/final_out`. `*remove_trRNA.bed` are needed for downstream analysis. Its format is:

```
chr1    3214161 3214286 ST-E00126:647:HL57HCCXY:1:2107:20354:17377      255     -
chr1    3539203 3539327 ST-E00126:647:HL57HCCXY:1:1202:30188:47562      255     -
chr1    3539203 3539328 ST-E00126:647:HL57HCCXY:1:2202:9820:48142       255     -
chr1    3540150 3540274 ST-E00126:647:HL57HCCXY:1:2105:32177:70592      255     -
chr1    3540150 3540274 ST-E00126:647:HL57HCCXY:1:1217:19380:22018      255     -
chr1    3612996 3613077 ST-E00126:647:HL57HCCXY:1:1117:10389:26484      255     -
```
This is a tab-delimited file. Each row represents a read.




# 3. Call peak and correction

We extract the reads on chr 3 for simplification:

```
grep "chr3" ~/scCAT_seq/five_pirme/final_out/O41_72_TKD180302275-N704-AK417_AHL57HCCXY_L1_2.fq_with_tag.fq.trimed.remainGGG_Aligned.out.sam_extract_uniquely_map.sam_extractmismatch_add_header_sorted_remove_trRNA.bed > ~/scCAT_seq/five_pirme/final_out/O41_72_TKD180302275-N704-AK417_AHL57HCCXY_L1_2.fq_with_tag.fq.trimed.remainGGG_Aligned.out.sam_extract_uniquely_map.sam_extractmismatch_add_header_sorted_remove_trRNA_chr3.bed

grep "chr3" ~/scCAT_seq/three_pirme/final_out/O41_72_TKD180302275-N704-AK417_AHL57HCCXY_L1_2.fq-common.out_withA10_remain_A5_Aligned.out.sam_extract_uniquely_map.sam_extractmismatch_add_header_sotred_remove_trRNA.bed > ~/scCAT_seq/five_pirme/final_out/O41_72_TKD180302275-N704-AK417_AHL57HCCXY_L1_2.fq_with_tag.fq.trimed.remainGGG_Aligned.out.sam_extract_uniquely_map.sam_extractmismatch_add_header_sorted_remove_trRNA_chr3.bed
```

* Features for peak correction:
1) TPM_of_peak: The total Tags per Million(TPM) value of the peak called by `CAGEr`.
2) TPM_of_Dominant_Site: The highest TPM value of all sites within a peak.
3) Gene_length: The length of the transcript annotated.
4) Peak_width: The width of the peak called by `CAGEr`.
5) Dominant_TPM_to_Smart2: The ratio of Dominant_TPM to the RPM value of the corresponding gene revealed by Smart-seq2.
6) Slope_smart2_curve: The slope of Smart-seq2 coverage curve around the peaks.
7) Trend_of_smart2_reads: Calculated by dividing the number of reads increased/decreased within 50bp distance by 50.
8) Percentage: The percentage of read counts of a peak to the total counts of a transcript.
9) Motif: Motif information around TSS/TES peak

* Correction:  
1) Predicted each individual peaks in the input file. The value `1` indicates a true TSS/TES peak while value `0` indicates a false TSS/TES peak





---

## 1. Call peak

We run:

```
cd call_peak
Rscript CAGE_dominant.R

for i in `ls|grep "_dominant_tes.bed"`
do
        bedtools intersect -s -a ${i} -b gencode_mm10_all_gene_genebody_and_downstream2k.bed -wa -wb > ${i%%.*}_genebody_downstream2k.bed
done

for i in `ls|grep "_dominant_tss.bed"`
do 
        bedtools intersect -s -a ${i} -b gencode_mm10_all_gene_ustream2k_and_genebody.bed -wa -wb > ${i%%.*}_upstream2k_and_genebody.bed
done 
```

`D44_52_3tail_dominant_tes.bed` and `D44_52_5cap_dominant_tss.bed` are generated. These file is peak calling output. `temp.RData` is used for next step.



## 2. Calculate TPM_of_peak, TPM_of_Dominant_Site, Peak_width and Gene_length

We run:

```
Rscript cal_slope_cattss.R D44_52_5cap_dominant_tss_upstream2k_and_genebody.bed D_merge_standard_smart_seq2.count
Rscript cal_slope_cattes.R D44_52_3tail_dominant_tes_genebody_downstream2k.bed D_merge_standard_smart_seq2.count
```

`tc_D44_52_5cap_peak.csv` and `tc_D44_52_3tail_peak.csv` are generated. In this step, we added some features for machine learning.



## 3. Calculate TPM_of_peak, TPM_of_Dominant_Site and Peak_width

We run:

```
Rscript cal_slope_cattss2.R tc_D44_52_5cap_peak.csv D_merge_standard_smart_seq2_sorted_chr3.wig
Rscript cal_slope_cattes2.R tc_D44_52_3tail_peak.csv  D_merge_standard_smart_seq2_sorted_chr3.wig
```

`tc_D44_52_5cap_peak.csv.csv` and `tc_D44_52_3tail_peak.csv.csv` are generated. In this step, we calculate the slope of Smart-seq2 coverage curve around the peaks and correlation.



## 4. Calculate percentage and Dominant_TPM_to_Smart2

We run:

```
Rscript cal_slope_cattss3.R tc_D44_52_5cap_peak.csv.csv
Rscript cal_slope_cattes3.R tc_D44_52_3tail_peak.csv.csv
```

`tc_D44_52_5cap_peak.csv.csv.csv` and `tc_D44_52_3tail_peak.csv.csv.csv` are generated. In this step, Percentage and Dominant_TPM_to_Smart2 are added.

## 5. Add Fantom5 and poly_DB

We run:

```
#### TSS
bedtools intersect -s -a temp_tss.bed -b tc_dsc.bed -wa -wb > temp_tss_in_FANTOM.bed
Rscript FANTOM.R tc_D44_52_5cap_peak.csv.csv.csv

#### TES
bedtools intersect -s -a temp_tes.bed -b mouse.PAS100_mm10.bed -wa -wb > temp_tes_in_polydb.bed
Rscript polydb.R tc_D44_52_3tail_peak.csv.csv.csv
```

`tc_D44_52_5cap_peak.csv.csv.csv.csv` and `tc_D44_52_3tail_peak.csv.csv.csv.csv` are generated. In this step, comparison to Fantom5 and poly_DB are added.

## 6. Add motif information

In this step, mm10.fa genome file is needed. 

We run:

```
#### Find TATA-box, BREu, BREd around TSS.
python find_motif_re_TSS.py ~/index/mm10/mm10.fa tc_D44_52_5cap_peak.csv.csv.csv.csv tc_D44_52_5cap_peak.csv.csv.csv.csv.csv

#### Find polyA singal around TES.
python find_motif_re_TES.py ~/index/mm10/mm10.fa tc_D44_52_3tail_peak.csv.csv.csv.csv tc_D44_52_3tail_peak.csv.csv.csv.csv.csv
```

In this step, `tc_D44_52_5cap_peak.csv.csv.csv.csv.csv` and `tc_D44_52_3tail_peak.csv.csv.csv.csv.csv` are generated. We search motif form upstream 40 bp to 0 bp, relatived to dominant TSS/TES position. `NA` represents there is no motif in this region. `Number` means motif is in this position. `;` means more than 1 motif in this area. 


## 7. Unfold motif and change gene length to mean transcript length

```
Rscript unfold_tss.R tc_D44_52_5cap_peak.csv.csv.csv.csv.csv
Rscript unfold_tes.R tc_D44_52_3tail_peak.csv.csv.csv.csv.csv
```

In this step, `tc_D44_52_5cap_peak.csv.csv.csv.csv.csv.csv` and `tc_D44_52_3tail_peak.csv.csv.csv.csv.csv.csv` are generated. The value `1` indicates that there is motif in this position while value `0` indicates no motif enrichment.

## 8. Prediction and correction

To be added

We run:

```
mv tc_D44_52_5cap_peak.csv.csv.csv.csv.csv.csv ../data/
mv tc_D44_52_3tail_peak.csv.csv.csv.csv.csv.csv ../data/
cd ../
python ./bin/Demo_DRG.py
```

The new folder named `output` will be built after completion of the pipeline, which contains all the output files, including the `\*.csv` file. 

The last column in the `CSV` files shows the predicted classification corresponding to each individual peaks in the input file. The value `1` indicates a true TSS/TES peak while value `0` indicates a false TSS/TES peak.











