# Code for data processing and peak calling
Code in this directory is used to process the data generated by scCAT-seq, C1 CAGE, BAT-seq, etc. Fastq file is needed and the output file format is "BED". Each Row represents read in corresponding position.

Then BED file as input is needed to call peak using CAGEr R package.

The workflow of data of scCAT-seq 5', C1 CAGE, C1 STRT, Arguel et al. are similar:
1) Find reads with TSO primer: Reads with TSO primer sequence at 5' are considered to further processing.
2) Trim TSO primer but retain GGG: TSO primer is trimmed. "GGG" was used to remove reads with strand invasion.
3) Mapping: Reads are aligned to mouse genome (mm10) or human genome (hg38). We only select uniquely mapped reads.
4) Split reads aligned to plus stand and minus strand: This step is to prepare for step 5.
5) Extract reads with mismatch at 5': The template-switching (TS) oligonucleotide may hybridize to the first strand cDNA due to sequence complementarity before the RT has finished polymerizing. Artifact that introduced by incomplete reverse transcription process is called strand invasion. We suppose that if "GGG" aligned, reads are strand invasion drivern artifacts. if "GGG" don't aligned, reads are derived from complete reverse transcription.
6) Convert SAM to BED: As BED format file can be used as input for CAGEr R package, we generate SAM to BED.
