

Code in this directory is used to process the data generated by scCAT-seq, C1 CAGE, BAT-seq, etc. `Fastq` file is needed and the output file format is `BED`. Each Row represents read in corresponding position.

Then `BED` file as input is needed to call peak using `CAGEr` R package.

Here is the scCAT-seq 5' data and 3' data processing workflow. To see detail imformation of other data processing, please see `C1_CAGE_5_data_processing.sh`, `C1_STRT_5_data_processing.sh`, `Arguel_et_al_5_data_processing.sh` and `BAT-seq_3_data_processing.sh`.


We have uploaded test data. Reader can download at [here](https://drive.google.com/open?id=1t8oLqAIWWy32yf5g3NOfKm10-i0pBITy) and [here](https://drive.google.com/open?id=1Z4xEVmkip3aq56Jp5k-0qBLmZ9oR-Lyk).

data processing and peak correction


# 1. Preparation

In this workflow, we assume that the genome fa file is located in `~/index/mm10/mm10.fa`. STAR index is located in `~/index/mm10_STAR/`. fastq files from scCAT-seq are located in `~/fastq/`.

STAR index must be prepared before running this workflow. For build index, you can run:

```
## you can change fa and gtf file path
STAR --runThreadN 24 --runMode genomeGenerate \
--genomeDir ~/index/mm10_STAR/ \
--genomeFastaFiles ~/index/mm10/mm10.fa \
--sjdbGTFfile ~/index/mm10/gencode.vM18.annotation.gtf \
--sjdbOverhang 150
```

# 2. Convert to bed

We use `convert_to_bed.sh` to process the data. The script usage is:

```
sh convert_to_bed.sh <genome fa file> <STAR index> <fastq file dir> <output dir>
```

For example:
```
sh convert_to_bed.sh ~/index/mm10/mm10.fa ~/index/mm10_STAR/ ~/fastq/ ~/scCAT_seq/
```

Output files are stored in `~/scCAT_seq/five_pirme/final_out` and `~/scCAT_seq/three_pirme/final_out`. `*remove_trRNA.bed` are needed for downstream analysis. Its format is:

```
chr1    3214161 3214286 ST-E00126:647:HL57HCCXY:1:2107:20354:17377      255     -
chr1    3539203 3539327 ST-E00126:647:HL57HCCXY:1:1202:30188:47562      255     -
chr1    3539203 3539328 ST-E00126:647:HL57HCCXY:1:2202:9820:48142       255     -
chr1    3540150 3540274 ST-E00126:647:HL57HCCXY:1:2105:32177:70592      255     -
chr1    3540150 3540274 ST-E00126:647:HL57HCCXY:1:1217:19380:22018      255     -
chr1    3612996 3613077 ST-E00126:647:HL57HCCXY:1:1117:10389:26484      255     -
```

This is a tab-delimited file. Each row represents a read.




# 3. Call peak and feature generation

In this step, we extract the reads on chr 3 for simplification.

Features for peak correction are listed below:

1) TPM_of_peak: The total Tags per Million(TPM) value of the peak called by `CAGEr`.
2) TPM_of_Dominant_Site: The highest TPM value of all sites within a peak.
3) Gene_length: The length of the transcript annotated.
4) Peak_width: The width of the peak called by `CAGEr`.
5) Dominant_TPM_to_Smart2: The ratio of Dominant_TPM to the RPM value of the corresponding gene revealed by Smart-seq2.
6) Slope_smart2_curve: The slope of Smart-seq2 coverage curve around the peaks.
7) Trend_of_smart2_reads: Calculated by dividing the number of reads increased/decreased within 50bp distance by 50.
8) Percentage: The percentage of read counts of a peak to the total counts of a transcript.
9) Motif: Motif information around TSS/TES peak


We use `callpeak_correction_TSS.sh` and `callpeak_correction_TES.sh` to generate features and compare peaks to `Gencode`, `FANTOM5` and `PolyA_DB` databases. The script usage is:

```
sh callpeak_correction_TSS.sh <bed file> <Smart-seq2 readcount> <Smart-seq2 wig file> <FANTOM5 file>
sh callpeak_correction_TES.sh <bed file> <Smart-seq2 readcount> <Smart-seq2 wig file> <PolyA_db file>
```

For example:

```
sh callpeak_correction_TSS.sh ~/scCAT_seq/five_pirme/final_out/O41_72_TKD180302275-N704-AK417_AHL57HCCXY_L1_2.fq_with_tag.fq.trimed.remainGGG_Aligned.out.sam_extract_uniquely_map.sam_extractmismatch_add_header_sorted_remove_trRNA.bed O_merge_standard_smart_seq2.count O_merge_standard_smart_seq2_sorted_chr3.wig tc_ovary.bed
sh callpeak_correction_TES.sh ~/scCAT_seq/three_pirme/final_out/O41_72_TKD180302275-N704-AK417_AHL57HCCXY_L1_1.fq-common.out_withA10_remain_A5_Aligned.out.sam_extract_uniquely_map.sam_extractmismatch_add_header_sorted_remove_trRNA.bed O_merge_standard_smart_seq2.count O_merge_standard_smart_seq2_sorted_chr3.wig mouse.PAS100_mm10.bed
```

`*final.csv` are ouput files. Its format is: 

```
chr,peak_start,peak_end,gene,strand,peak_RPM,dominant_tss_position,dominant_tss_RPM,gene_length,peak_length,dominant_tss_RPM/smart_seq2_RPM,slope,correlation,percentage,annotated_peak,isindsc,BREu_motif_minus_49,BREu_motif_minus_48,BREu_motif_minus_47,BREu_motif_minus_46,BREu_motif_minus_45,BREu_motif_minus_44,BREu_motif_minus_43,BREu_motif_minus_42,BREu_motif_minus_41,BREu_motif_minus_40,BREu_motif_minus_39,BREu_motif_minus_38,BREu_motif_minus_37,BREu_motif_minus_36,BREu_motif_minus_35,BREu_motif_minus_34,BREu_motif_minus_33,BREu_motif_minus_32,BREu_motif_minus_31,BREu_motif_minus_30,BREu_motif_minus_29,BREu_motif_minus_28,BREu_motif_minus_27,BREu_motif_minus_26,BREu_motif_minus_25,BREu_motif_minus_24,BREu_motif_minus_23,BREu_motif_minus_22,BREu_motif_minus_21,BREu_motif_minus_20,BREu_motif_minus_19,BREu_motif_minus_18,BREu_motif_minus_17,BREu_motif_minus_16,BREu_motif_minus_15,BREu_motif_minus_14,BREu_motif_minus_13,BREu_motif_minus_12,BREu_motif_minus_11,BREu_motif_minus_10,BREu_motif_minus_9,BREu_motif_minus_8,BREu_motif_minus_7,BREu_motif_minus_6,BREu_motif_minus_5,BREu_motif_minus_4,BREu_motif_minus_3,BREu_motif_minus_2,BREu_motif_minus_1,BREu_motif_minus_0,BREd_motif_minus_49,BREd_motif_minus_48,BREd_motif_minus_47,BREd_motif_minus_46,BREd_motif_minus_45,BREd_motif_minus_44,BREd_motif_minus_43,BREd_motif_minus_42,BREd_motif_minus_41,BREd_motif_minus_40,BREd_motif_minus_39,BREd_motif_minus_38,BREd_motif_minus_37,BREd_motif_minus_36,BREd_motif_minus_35,BREd_motif_minus_34,BREd_motif_minus_33,BREd_motif_minus_32,BREd_motif_minus_31,BREd_motif_minus_30,BREd_motif_minus_29,BREd_motif_minus_28,BREd_motif_minus_27,BREd_motif_minus_26,BREd_motif_minus_25,BREd_motif_minus_24,BREd_motif_minus_23,BREd_motif_minus_22,BREd_motif_minus_21,BREd_motif_minus_20,BREd_motif_minus_19,BREd_motif_minus_18,BREd_motif_minus_17,BREd_motif_minus_16,BREd_motif_minus_15,BREd_motif_minus_14,BREd_motif_minus_13,BREd_motif_minus_12,BREd_motif_minus_11,BREd_motif_minus_10,BREd_motif_minus_9,BREd_motif_minus_8,BREd_motif_minus_7,BREd_motif_minus_6,BREd_motif_minus_5,BREd_motif_minus_4,BREd_motif_minus_3,BREd_motif_minus_2,BREd_motif_minus_1,BREd_motif_minus_0,TATA_motif_minus_49,TATA_motif_minus_48,TATA_motif_minus_47,TATA_motif_minus_46,TATA_motif_minus_45,TATA_motif_minus_44,TATA_motif_minus_43,TATA_motif_minus_42,TATA_motif_minus_41,TATA_motif_minus_40,TATA_motif_minus_39,TATA_motif_minus_38,TATA_motif_minus_37,TATA_motif_minus_36,TATA_motif_minus_35,TATA_motif_minus_34,TATA_motif_minus_33,TATA_motif_minus_32,TATA_motif_minus_31,TATA_motif_minus_30,TATA_motif_minus_29,TATA_motif_minus_28,TATA_motif_minus_27,TATA_motif_minus_26,TATA_motif_minus_25,TATA_motif_minus_24,TATA_motif_minus_23,TATA_motif_minus_22,TATA_motif_minus_21,TATA_motif_minus_20,TATA_motif_minus_19,TATA_motif_minus_18,TATA_motif_minus_17,TATA_motif_minus_16,TATA_motif_minus_15,TATA_motif_minus_14,TATA_motif_minus_13,TATA_motif_minus_12,TATA_motif_minus_11,TATA_motif_minus_10,TATA_motif_minus_9,TATA_motif_minus_8,TATA_motif_minus_7,TATA_motif_minus_6,TATA_motif_minus_5,TATA_motif_minus_4,TATA_motif_minus_3,TATA_motif_minus_2,TATA_motif_minus_1,TATA_motif_minus_0
chr3,4305637,4305664,Gm23976,-,5.73578014066427,4305664,1.14715602813285,101,27,4.22406595688299,0.26,0.939962341203223,0.625,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
chr3,4305774,4305775,Gm23976,-,1.14715602813285,4305775,1.14715602813285,101,1,4.22406595688299,0,0.325533064436891,0.125,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,1,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
chr3,4306611,4306612,Gm23976,-,2.29431205626571,4306612,2.29431205626571,101,1,8.44813191376599,0.02,0.859939415493515,0.25,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
chr3,5181890,5181909,RP24-257J5.1,-,5.73578014066427,5181909,2.29431205626571,677,19,3.07204796864218,0.1,0.919697490709145,0.833333333333333,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,1,0,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
chr3,5181950,5181951,RP24-257J5.1,-,1.14715602813285,5181951,1.14715602813285,677,1,1.53602398432109,0.1,0.923050831094628,0.166666666666667,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,0,0,0,0,0,0,0,1,1,1,0,0,0,0,0,0,0,1,1,1,1,0,0,0,0,0,0,0,1,1,0,0,1,1,0,1,0,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
```

This is a comma-delimited file. Each row represents a peak.

# 4. Peaks correction

To be added

* Correction:  
1) Predicted each individual peaks in the input file. The value `1` indicates a true TSS/TES peak while value `0` indicates a false TSS/TES peak


