

Code in this directory is used to process the data generated by scCAT-seq, C1 CAGE, BAT-seq, etc. Fastq file is needed and the output file format is "BED". Each Row represents read in corresponding position.

Then BED file as input is needed to call peak using CAGEr R package.

---

# Data processing for 5' data

The workflows of data of scCAT-seq 5', C1 CAGE, C1 STRT and Arguel et al. are similar. Here is the scCAT-seq 5' data processing workflow. To see detail imformation of other data processing, please see C1_CAGE_5_data_processing.sh, C1_STRT_5_data_processing.sh and Arguel_et_al_5_data_processing.sh.

## Prepare

Before process the data, we bulid some directory and move the script to "script_and_log" directory: 

```
mkdir ~/zjw/20190109
mkdir ~/zjw/20190109/5cap_read_with_tag
mkdir ~/zjw/20190109/trim_GTGGTATCAACGCAGAGTACAT
mkdir ~/zjw/20190109/mapping_output
mkdir ~/zjw/20190109/extract_uniquely_map
mkdir ~/zjw/20190109/split_plus_minus
mkdir ~/zjw/20190109/extract_mismatch
mkdir ~/zjw/20190109/final_out
mkdir ~/zjw/20190109/script_and_log
mv extractmismatch_plus_5'.py ~/zjw/20190109/script_and_log
mv extractmismatch_plus_3'.py ~/zjw/20190109/script_and_log
cd ~/zjw/20190109/script_and_log
```

Please make sure that fq files are list in ~/zjw/fastq_5cap_2018ab. STAR index must be prepared before running this workflow.

## Find reads with TSO primer

Reads with TSO primer sequence at 5' are considered to further processing. TSO primer in scCAT-seq data is `GTGGTATCAACGCAGAGTACATGGG`.

```
for i in `ls ~/zjw/fastq_5cap_2018ab/`
do
        cat ~/zjw/fastq_5cap_2018ab/${i} | paste - - - - | grep $'\t'"${a}" | awk -v FS="\t" -v OFS="\n" '{print $1, $2, $3, $4}' > ~/zjw/20190109/5cap_read_with_tag/${i}_with_tag.fq
done
```

Output files are stored in `~/zjw/20190109/5cap_read_with_tag/`.

## Trim TSO primer but retain GGG

To trim TSO primer, we run:

```
for i in `ls ~/zjw/20190109/5cap_read_with_tag`
do
        cutadapt -g GTGGTATCAACGCAGAGTACAT -o ~/zjw/20190109/trim_GTGGTATCAACGCAGAGTACAT/${i}.trimed.remainGGG ~/zjw/20190109/5cap_read_with_tag/${i}
done
```

In this step, we trim TSO primer. However `GGG` at the end of TSO primer was retained for further filter. 

Output files are stored in `~/zjw/20190109/trim_GTGGTATCAACGCAGAGTACAT/`.

## Alignment

For alignment, we run:

```
for i in `ls ~/zjw/20190109/trim_GTGGTATCAACGCAGAGTACAT/`
do
        STAR --runThreadN 24 --genomeDir ~/index/mm10_ERCC92trimpolyA_STAR/ --genomeLoad LoadAndKeep --readFilesIn ~/zjw/20190109/trim_GTGGTATCAACGCAGAGTACAT/${i} --outFileNamePrefix ~/zjw/20190109/mapping_output/${i}_ --outSAMtype SAM --outFilterMultimapNmax 1 --outFilterScoreMinOverLread 0.6 --outFilterMatchNminOverLread 0.6
done
```

Output files are stored in `~/zjw/20190109/mapping_output/`

## Extract uniquely mapped reads

We run:

```
for i in `ls ~/zjw/20190109/mapping_output|grep "sam"`
do
        samtools view ~/zjw/20190109/mapping_output/${i} |grep 'NH:i:1'$'\t''' > ~/zjw/20190109/extract_uniquely_map/${i}_extract_uniquely_map.sam
done
```

Output files are stored in `~/zjw/20190109/extract_uniquely_map/`

## Split reads aligned to plus stand and minus strand

For further filter, we run:

```
for i in `ls ~/zjw/20190109/extract_uniquely_map | grep "sam"`
do
        cat ~/zjw/20190109/extract_uniquely_map/${i} | awk '{FS=" "}{if ($2==0 || $2==256){print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7"\t"$8"\t"$9"\t"$10"\t"$11"\t"$12"\t"$13"\t"$14"\t"$15}}' >                         ~/zjw/20190109/split_plus_minus/${i}_plus
        cat ~/zjw/20190109/extract_uniquely_map/${i} | awk '{FS=" "}{if ($2==16 || $2==272){print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"$7"\t"$8"\t"$9"\t"$10"\t"$11"\t"$12"\t"$13"\t"$14"\t"$15}}' > ~/zjw/20190109/split_plus_minus/${i}_minus
done
```

Output files are stored in `~/zjw/20190109/split_plus_minus/`

## Extract reads with mismatch at 5'

We run:

```
for i in `ls ~/zjw/20190109/split_plus_minus | grep "extract_uniquely_map.sam_plus"`
do
        python ~/zjw/20190109/script_and_log/extractmismatch_plus_5'.py -i ~/zjw/20190109/split_plus_minus/${i} -o ~/zjw/20190109/extract_mismatch/${i}_extractmismatch
        python ~/zjw/20190109/script_and_log/extractmismatch_minus_5'.py -i ~/zjw/20190109/split_plus_minus/${i%_*}_minus -o ~/zjw/20190109/extract_mismatch/${i%_*}_minus_extractmismatch
        cat ~/zjw/20190109/extract_mismatch/${i}_extractmismatch ~/zjw/20190109/extract_mismatch/${i%_*}_minus_extractmismatch > ~/zjw/20190109/extract_mismatch/${i%_*}_extractmismatch
done
```

The template-switching (TS) oligonucleotide may hybridize to the first strand cDNA due to sequence complementarity before the RT has finished polymerizing. Artifact that introduced by incomplete reverse transcription process is called strand invasion. We suppose that if "GGG" aligned, reads are strand invasion drivern artifacts. If "GGG" don't aligned, reads are derived from complete reverse transcription.

Output files are stored in `~/zjw/20190109/extract_mismatch/`

## Convert SAM to BED

BED files are needed for CAGEr, so we run:

```
for i in `ls ~/zjw/20190109/extract_mismatch | grep "sam_extractmismatch"`
do
        samtools view -b -T ~/index/mm10_ERCC92/mm10_ERCC92trimpolyA.fa ~/zjw/20190109/extract_mismatch/${i} | samtools view -b >  ~/zjw/20190109/final_out/${i}_add_header.bam
        samtools sort ~/zjw/20190109/final_out/${i}_add_header.bam -o ~/zjw/20190109/final_out/${i}_add_header_sorted.bam
        samtools index ~/zjw/20190109/final_out/${i}_add_header_sorted.bam
        bedtools bamtobed -i ~/zjw/20190109/final_out/${i}_add_header_sorted.bam > ~/zjw/20190109/final_out/${i}_add_header_sorted.bed
done
```

Output files are stored in `~/zjw/20190109/final_out/`


## Remove useless end

As the library is pair-end reads, we remove one side which doesn't contain TSS information.

```
for i in `ls  ~/zjw/20190109/final_out |grep "bed"|grep "L1_1"`
do
        a=$(wc -l ~/zjw/20190109/final_out/${i}|awk '{print $1}')
        b=$(wc -l ~/zjw/20190109/final_out/${i%%L1_1*}L1_2.fq_with_tag.fq.trimed.remainGGG_Aligned.out.sam_extract_uniquely_map.sam_extractmismatch_add_header_sorted.bed|awk '{print $1}')
        echo ${a}
        echo ${b}
        if [ ${a} -gt ${b} ]; then
                rm ~/zjw/20190109/final_out/${i%%L1_1*}L1_2*
        else
                rm ~/zjw/20190109/final_out/${i%%L1_1*}L1_1*
        fi
done
```

1) Find reads with TSO primer: Reads with TSO primer sequence at 5' are considered to further processing.
2) Trim TSO primer but retain GGG: TSO primer is trimmed. "GGG" was used to remove reads with strand invasion.
3) Mapping: Reads are aligned to mouse genome (mm10) or human genome (hg38). We only select uniquely mapped reads.
4) Split reads aligned to plus stand and minus strand: This step is to prepare for step 5.
5) Extract reads with mismatch at 5': The template-switching (TS) oligonucleotide may hybridize to the first strand cDNA due to sequence complementarity before the RT has finished polymerizing. Artifact that introduced by incomplete reverse transcription process is called strand invasion. We suppose that if "GGG" aligned, reads are strand invasion drivern artifacts. If "GGG" don't aligned, reads are derived from complete reverse transcription.
6) Convert SAM to BED: As BED format file can be used as input for CAGEr R package, we generate SAM to BED.
7) Generate CAGEset object in R: See R script generate_CAGEset_5'.R 

---

The workflows of data of scCAT-seq 3' and BAT-seq are similar. Here is the scCAT-seq 5' data processing workflow. To see detail imformation of BAT-seq data processing, please see BAT-seq_3_data_processing.sh.

1) Find reads with oligo(dT) primer: Reads with oligo(dT) primer sequence at 5'. We define reads with oligo(dT) primer sequence at 5' as R1 reads.
2) Find R2 reads: Perl script cmpfastq_pe.pl is used to find R2 reads which its corresponding R1 reads with oligo(dT) primes.
3) Trim A10 but retain A5 at R2 reads: Poly A sequences in the end of R2 were further trimmed with 5 A bases left at the 3â€™ side. "AAAAA" was used to remove reads with internal priming.
4) Mapping: Reads are aligned to mouse genome (mm10). We only select uniquely mapped reads.
5) Split reads aligned to plus stand and minus strand: This step is to prepare for step 6.
6) Extract reads with mismatch at 3': Oligo(dT) primers can also anneal to internal A-rich sequences, a phenomenon called internal priming, leading to the generation of artifact. If "AAAAA" aligned, reads are internal priming drivern artifacts. If "GGG" don't aligned, reads are supposed to contain true polyA site.
7) Convert SAM to BED: As BED format file can be used as input for CAGEr R package, we generate SAM to BED.
8) Generate CAGEset object in R: See R script generate_CAGEset_3'.R

